# Smart ML System for Predictive Analysis of Student Performance

## 02 - Data Cleaning

This notebook focuses on cleaning and preparing the student performance dataset for exploratory data analysis, feature engineering, and machine learning.

The cleaning process includes:
- Checking and correcting data types
- Standardizing column names
- Checking missing and duplicate records
- Identifying invalid values
- Investigating derived variables
- Preventing data leakage
- Preparing the cleaned dataset for further analysis

In [1]:
import pandas as pd
import numpy as np

from pathlib import Path

In [2]:
file_path = Path("../data/raw/student_performance_analysis.xlsx")

df = pd.read_excel(
    file_path,
    sheet_name="Raw_Data"
)

df.head()

,student_id,weekly_self_study_hours,attendance_percentage,class_participation,total_score,grade,Perfomance Category,attendence category,Study hours category,Participation level
0,1,18.5,95.6,3.8,97.9,A,Excellent,High,Medium Study,Low
1,2,14.0,80.0,2.5,83.9,B,Good,Medium,Medium Study,Low
2,3,19.5,86.3,5.3,100.0,A,Excellent,Medium,Medium Study,Medium
3,4,25.7,70.2,7.0,100.0,A,Excellent,Low,High Study,High
4,5,13.4,81.9,6.9,92.0,A,Excellent,Medium,Medium Study,Medium


In [3]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

df.columns

Index(['student_id', 'weekly_self_study_hours', 'attendance_percentage',
       'class_participation', 'total_score', 'grade', 'perfomance_category',
       'attendence_category', 'study_hours_category', 'participation_level'],
      dtype='str')

In [4]:
df = df.rename(columns={
    "perfomance_category": "performance_category",
    "attendence_category": "attendance_category"
})

df.columns

Index(['student_id', 'weekly_self_study_hours', 'attendance_percentage',
       'class_participation', 'total_score', 'grade', 'performance_category',
       'attendance_category', 'study_hours_category', 'participation_level'],
      dtype='str')

In [5]:
df.dtypes

student_id                   int64
weekly_self_study_hours    float64
attendance_percentage      float64
class_participation        float64
total_score                float64
grade                          str
performance_category           str
attendance_category            str
study_hours_category           str
participation_level            str
dtype: object

In [6]:
df.isnull().sum()

student_id                 0
weekly_self_study_hours    0
attendance_percentage      0
class_participation        0
total_score                0
grade                      0
performance_category       0
attendance_category        0
study_hours_category       0
participation_level        0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
print("Negative study hours:")
print((df["weekly_self_study_hours"] < 0).sum())

print("\nInvalid attendance:")
print(((df["attendance_percentage"] < 0) | (df["attendance_percentage"] > 100)).sum())

print("\nInvalid participation:")
print(((df["class_participation"] < 0) | (df["class_participation"] > 100)).sum())

print("\nInvalid total scores:")
print(((df["total_score"] < 0) | (df["total_score"] > 100)).sum())

Negative study hours:
0

Invalid attendance:
0

Invalid participation:
0

Invalid total scores:
0


In [9]:
pd.crosstab(
    df["grade"],
    pd.cut(
        df["total_score"],
        bins=[0, 50, 60, 70, 80, 90, 100],
        include_lowest=True
    )
)

total_score,"(-0.001, 50.0]","(50.0, 60.0]","(60.0, 70.0]","(70.0, 80.0]","(80.0, 90.0]","(90.0, 100.0]"
grade,,,,,,
A,0,0,0,0,508,2297
B,0,0,4,807,470,0
C,0,128,532,0,0,0
D,110,122,0,0,0,0
F,32,0,0,0,0,0


In [10]:
pd.crosstab(
    df["performance_category"],
    pd.cut(
        df["total_score"],
        bins=[0, 50, 60, 70, 80, 90, 100],
        include_lowest=True
    )
)

total_score,"(-0.001, 50.0]","(50.0, 60.0]","(60.0, 70.0]","(70.0, 80.0]","(80.0, 90.0]","(90.0, 100.0]"
performance_category,,,,,,
Average,0,4,536,344,0,0
Excellent,0,0,0,0,12,2297
Good,0,0,0,463,966,0
Needs Improvement,142,246,0,0,0,0


In [11]:
df.groupby("study_hours_category")["weekly_self_study_hours"].agg(
    ["min", "max", "mean", "count"]
)

,min,max,mean,count
study_hours_category,,,,
High Study,20.0,40.0,24.097880,1179
Low Study,0.0,9.9,6.026139,1163
Medium Study,10.0,19.9,15.033996,2668


In [12]:
df.groupby("attendance_category")["attendance_percentage"].agg(
    ["min", "max", "mean", "count"]
)

,min,max,mean,count
attendance_category,,,,
High,90.0,100.0,95.448018,1564
Low,50.0,74.9,69.919849,796
Medium,75.0,89.9,82.716830,2650


In [13]:
df.groupby("participation_level")["class_participation"].agg(
    ["min", "max", "mean", "count"]
)

,min,max,mean,count
participation_level,,,,
High,7.0,10.0,8.223722,1585
Low,0.0,3.9,2.871214,733
Medium,4.0,6.9,5.528455,2692


In [14]:
model_df = df[
    [
        "weekly_self_study_hours",
        "attendance_percentage",
        "class_participation",
        "total_score"
    ]
].copy()

model_df.head()

,weekly_self_study_hours,attendance_percentage,class_participation,total_score
0,18.5,95.6,3.8,97.9
1,14.0,80.0,2.5,83.9
2,19.5,86.3,5.3,100.0
3,25.7,70.2,7.0,100.0
4,13.4,81.9,6.9,92.0


In [15]:
model_df.shape

(5010, 4)

In [16]:
model_df.isnull().sum()

weekly_self_study_hours    0
attendance_percentage      0
class_participation        0
total_score                0
dtype: int64

In [17]:
model_df.duplicated().sum()

np.int64(1)

In [18]:
model_df[model_df.duplicated(keep=False)]

,weekly_self_study_hours,attendance_percentage,class_participation,total_score
641,26.7,100.0,6.0,100.0
2404,26.7,100.0,6.0,100.0


In [19]:
processed_path = Path("../data/processed")
processed_path.mkdir(parents=True, exist_ok=True)

model_df.to_csv(
    processed_path / "student_performance_model_data.csv",
    index=False
)

print("Modeling dataset saved successfully.")

Modeling dataset saved successfully.


## Data Cleaning Summary

The dataset was successfully cleaned and prepared for further analysis.

### Data Quality
- Total records: 5,010
- Missing values: None
- Original duplicate records: None
- Invalid numerical values: None

### Data Leakage Investigation
The following categorical variables were identified as derived variables:

- `grade` → derived from `total_score`
- `performance_category` → derived from `total_score`
- `study_hours_category` → derived from `weekly_self_study_hours`
- `attendance_category` → derived from `attendance_percentage`
- `participation_level` → derived from `class_participation`

These variables are retained in the original dataframe for descriptive analysis but are excluded from the initial machine learning feature set to prevent target leakage and redundant representation of numerical variables.

### Modeling Dataset

The initial regression model will use:

**Features**
- `weekly_self_study_hours`
- `attendance_percentage`
- `class_participation`

**Target**
- `total_score`

The modeling dataset has been saved as:

`data/processed/student_performance_model_data.csv`

This dataset will be used for exploratory data analysis, feature engineering, model training, and evaluation.